google/cloud/sql/connector/rate_limiter.py

```diff
@@ -0,0 +1,114 @@
+        Default: 0
+
+    :type loop: asyncio.AbstractEventLoop
+    :param: loop:
+        The event loop to use. If not provided, the default event loop will be used.
+
+
+    """
+
+    def __init__(
+        self,
+        burst_size: int = 1,
+        interval: float = 60,
+        loop: asyncio.AbstractEventLoop = None,
+    ) -> None:
+        self.interval = interval
+        self.burst_size = burst_size
+        self._loop = loop or asyncio.get_event_loop()
+        self._tokens = asyncio.BoundedSemaphore(burst_size, loop=self._loop)
+        self._last_token_update = time.time()
+
+    def _add_tokens(self) -> None:
+        """
+        Calculates how much time has passed since the last leak and removes the
+        appropriate amount of events from the queue.
+        Leaking is done lazily, meaning that if there is a large time gap between
+        leaks, the next set of calls might be a burst if burst_size > 1
+        """
+        time_elapsed = time.time() - self._last_token_update
+        amount = math.floor(time_elapsed * self.burst_size / self.interval)
+        # don't allow more than burst_size to leak at once
+        leak_amount = min(amount, self.burst_size)
+        if amount > 0:
+            self._last_token_update = time.time()
+        for _ in range(leak_amount):
+            try:
+                self._tokens.release()
+            except ValueError:
+                # we want to keep adding tokens until the semaphore reaches its upper bound
+                break
+
+    def _token_acquired(self, waiter_task: asyncio.Task) -> bool:
+        """
+        Releases tasks out of the queue if enough time has elapsed.
+        """
+        self._add_tokens()
+        return waiter_task.done()
+
+    async def acquire(self) -> None:
+        """
+        Adds an event to the queue and creates a waiter task to wait
+        for the event's turn in the queue.
+        """
+        waiter_task = self._loop.create_task(self._tokens.acquire())
+        while not self._token_acquired(waiter_task):
+            try:
+                # await until enough time has passed that another event is released from the queue. If an event is released before this time, the waiter_task will complete and we will break out of the loop.
+                await asyncio.wait_for(
+                    asyncio.shield(waiter_task), timeout=self.interval / self.burst_size
+                )
+            except asyncio.TimeoutError:
+                # allow for another call to self._release_from_queue() which will remove and set events from the queue
+                pass
+
+    async def __aenter__(self) -> None:
+        await self.acquire()
+        return None
+
+    async def __aexit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc: Optional[BaseException],
+        tb: Optional[TracebackType],
+    ) -> None:
+        return None
```

Comment:
it's neat you added these here, but are you actually using them?

Reply:
I'm using a the rate limiter as a context manager in the tests. I was using the context manager in `perform_refresh` in an earlier draft, but switched it to just using an acquire. I could delete these and change the tests if you want.

Reply:
yeah I think we should update the tests to use the same method.